flink-core/src/main/java/org/apache/flink/api/common/state/ListState.java

```diff
@@ -33,4 +35,14 @@
  * @param <T> Type of values that this list state keeps.
  */
 @PublicEvolving
-public interface ListState<T> extends MergingState<T, Iterable<T>> {}
+public interface ListState<T> extends MergingState<T, Iterable<T>> {
+	/**
+	 * Updates the state of the current key for the given source namespaces into the state of
+	 * the target namespace.
+	 *
+	 * @param values The target namespace where the merged state should be stored.
+	 *
+	 * @throws Exception The method may forward exception thrown internally (by I/O or functions).
+	 */
+	void update(List<T> values) throws Exception;
```

Comment:
We should define and document what happens if this is called with `null` or empty list. Does this lead to different results for `get()` ?

Reply:
Let me document it. They are of the same result - return `null` when given an empty list

Reply:
Ok, if the outcome for both cases should be the same, wouldn't it be nicer to always return the empty iterator and avoid `null`s if we can? This might be a small thing but safe users from debugging some NPEs :-)

Reply:
@StefanRRichter agree. The correct way to do should be changing `ListState#get()`. I filed [FLINK-8364](https://issues.apache.org/jira/browse/FLINK-8364) and will work on it after this PR is checked in